t/op/join.t

```diff
@@ -128,3 +128,84 @@ package o { use overload q|""| => sub { ${$_[0]}++ } }
 for(1,2) { push @_, \join "x", 1 }
 isnt $_[1], $_[0],
     'join(const, const) still returns a new scalar each time';
+
+# tests from GH #21458
+# simple tied variable
+{
+    package S;
+    our $fetched;
+    sub TIESCALAR { my $x = '-';   $fetched = 0; bless \$x }
+    sub FETCH     { my $y = shift; $fetched++;   $$y }
+
+    package main;
+    my $t;
+
+    tie $t, 'S';
+    is( join( $t, a .. c ), 'a-b-c', 'tied separator' );
+    is( $S::fetched,    1,       'FETCH called once' );
+
+    tie $t, 'S';
+    is( join( $t, 'a' ), 'a', 'tied separator on single item join' );
+    is( $S::fetched,     0,   'FETCH not called' );
+
+    tie $t, 'S';
+    is( join( $t, 'a', $t, 'b', $t, 'c' ),
+        'a---b---c', 'tied separator also in the join arguments' );
+    is( $S::fetched, 3, 'FETCH called 1 + 2 times' );
+}
+# self-modifying tied variable
+{
+
+    package SM;
+    our $fetched;
+    sub TIESCALAR { my $x = "1";   $fetched = 0; bless \$x }
+    sub FETCH     { my $y = shift; $fetched++;   $$y += 3 }
+
+    package main;
+    my $t;
+
+    tie $t, "SM";
+    is( join( $t, a .. c ), 'a4b4c', 'tied separator' );
+    is( $SM::fetched,   1,       'FETCH called once' );
+
+    tie $t, "SM";
+    is( join( $t, 'a' ), 'a', 'tied separator on single item join' );
+    is( $SM::fetched,    0,   'FETCH not called' );
+
+    tie $t, "SM";
+    is( join( $t, "a", $t, "b", $t, "c" ),
+        'a474b4104c', 'tied separator also in the join arguments' );
+    is( $SM::fetched, 3, 'FETCH called 1 + 2 times' );
+}
+{
+    # see GH #21484
+    my $expect = "a\x{100}x\x{100}b\n";
+    utf8::encode($expect);
+    fresh_perl_is(<<'CODE', $expect, {}, "modifications delim from magic should be ignored");
+my $n = 1;
```

Comment:
I'm not quite sure what `$n` is doing in this test. It is only ever 1 and only ever used as `STR x $n`. It could probably be removed for simplicity, unless it is being used to defeat some const-folding..? But it doesn't appear so to me.

Reply:
I've added an explanation of why I'm doing the `x $n` in each case, and yes, it's being done to prevent constant folding, or at least prevent CoW between `$sep` and some SVOP/PADOP.

If you wondered about it then and explanation can save someone else from having to do code archaeology trying to figure out the intent of the text.

Reply:
Ahah, yes that makes sense. A good comment for some otherwise-subtle logic.